workflow/controller/workflowpod.go

```diff
@@ -320,9 +297,19 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
 		}
 	}
 
+	deadline := time.Time{}
+	if woc.workflowDeadline != nil {
+		deadline = *woc.workflowDeadline
+	}
+	if deadline.IsZero() || opts.executionDeadline.Before(deadline) {
+		deadline = opts.executionDeadline
+	}
```

Comment:
could you put this into its own func and add some tests?

Reply:
Yes. I also found an issue: the second if should check that `!opts.executionDeadline.IsZero()`, or else it sets the deadline back to zero even if the `woc.workflowDeadline` is valid.